tla-types/src/main/scala/at/forsyte/apalache/tla/typedbuilder/TT1OperatorSignatureMatcher.scala

```diff
@@ -0,0 +1,819 @@
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T,T_1,...,T_n . (T, T_1, Set(T_1),..., T_m, Set(T_m)) => Set(T)
+  def matchSetMapSig(args: Seq[TlaEx]): SignatureMatch = {
+    val nArgs = args.length
+    assert(nArgs >= 3 && nArgs % 2 == 1)
+    val n = (nArgs - 1) / 2
+    val t = typeVarGenerator.getUnique
+    val ts = typeVarGenerator.getNUnique(n)
+    val allPairs = ts flatMap { x => Seq(x, SetT1(x)) }
+    val operT =
+      OperT1(
+          t +: allPairs,
+          SetT1(t)
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(T)) => Set(Set(T))
+  def matchPowSetSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(t)),
+          SetT1(SetT1(t))
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(Set(T))) => Set(T)
+  def matchBigUnionSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(SetT1(t))),
+          SetT1(t)
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(T)) => Bool
+  def matchSetPredicateSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(t)),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Set(T)) => Int
+  def matchCardinalitySig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(SetT1(t)),
+          IntT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (T) => T
+  def matchEndomorphismSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(t),
+          t
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (Bool,T) => Bool
+  def matchStutterSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(BoolT1(), t),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T . (T,Bool) => Bool
+  def matchTemporalWithVarSig(args: Seq[TlaEx]): SignatureMatch = {
+    val t = typeVarGenerator.getUnique
+    val operT =
+      OperT1(
+          Seq(t, BoolT1()),
+          BoolT1()
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T_1,T_2 . (Set(T_1), Set(T_2)) => Set(T_1 -> T2)
+  def matchFunSetSig(args: Seq[TlaEx]): SignatureMatch = {
+    val Seq(t1, t2) = typeVarGenerator.getNUnique(2)
+    val operT =
+      OperT1(
+          Seq(SetT1(t1), SetT1(t2)),
+          SetT1(FunT1(t1, t2))
+      )
+    matchArgImplied(operT, args)
+  }
+
+  // \A T,T_1,...,T_n . (T, T_1, Set(T_1),..., T_m, Set(T_m)) =>
+  // (case n = 1) : T_1 -> T
+  // (case n > 1) : (T_1, ..., T_n) -> T
+  def matchFunDefSig(args: Seq[TlaEx]): SignatureMatch = {
+    val nArgs = args.length
+    assert(nArgs >= 3 && nArgs % 2 == 1)
```

Comment:
This assert is the condition of `TlaFunOper.funDef.arity`, which takes an odd number of args (but not 1). Basically, whenever you have a built-in variadic operator, you want to check that you have the correct pairity (the minimal number is likely a redundant part of this assert, because the builder interface can guarantee the minimal number of arguments, but not the pairity). I do agree that it might be a good idea to throw the same kind of exception that `OperEx` construction does (`require(oper.isCorrectArity(args.size), [msg])`).